### PR TITLE
fix: .tsbuildinfo is not written in subsequent builds

### LINKED ISSUE #186: Project references fail to build when not clean
It seems like project references can only build from a clean workspace. Any non-exact future builds will fail

Error:
```
bazel build //examples/project_references/lib_a:lib_a

INFO: Build option --action_env has changed, discarding analysis cache.
INFO: Analyzed target //examples/project_references/lib_a:lib_a (2 packages loaded, 332 targets configured).
INFO: Found 1 target...
ERROR: /Users/ray/workspace/rules_ts/examples/project_references/lib_a/BUILD.bazel:15:11: output 'examples/project_references/lib_a/lib_a.tsbuildinfo' was not created
ERROR: /Users/ray/workspace/rules_ts/examples/project_references/lib_a/BUILD.bazel:15:11: Compiling TypeScript project //examples/project_references/lib_a:lib_a [tsc -p examples/project_references/lib_a/tsconfig.json] failed: not all outputs were created or valid
Target //examples/project_references/lib_a:lib_a failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1.393s, Critical Path: 0.25s
INFO: 18 processes: 8 internal, 9 local, 1 worker.
FAILED: Build did NOT complete successfully
```

Repro steps:

`git clone https://github.com/aspect-build/rules_ts`
`cd rules_ts`
`bazel build //examples/project_references/lib_a:lib_a # works`
`bazel build //examples/project_references/lib_a:lib_a # works because caches are exact`
`bazel build --action_env="avariable=1" //examples/project_references/lib_a:lib_a # force an unexact rebuild. Fails`
`bazel build //examples/project_references/lib_a:lib_a # will now always fail too`
`bazel clean`
`bazel build //examples/project_references/lib_a:lib_a # works again`

 --verbose_failures shows no additional debug info

I'm using an M1 mac if that matters. I haven't ruled out that something else is wrong with my set up. Happy to gather more debug info if there's methods other than verbose_failures

### PR DESCRIPTION
**Pull request by thesayyn**
fixes #186 #128

### CODE REVIEW

File: ts/private/ts_project_worker.js [hunk 1]
```diff
@@ -279,32 +279,36 @@ function createEmitAndLibCacheAndDiagnosticsProgram(
     );
 
     /** Emit Cache */
-    /** @type {Map} */
-    const emittedFiles = (host.emittedFiles = host.emittedFiles || new Map());
-    /** @type {Map} */
-    const emittingMap = (host.emittedFilesWeak = host.emittedFilesWeak || new Map());
+    const NOT_FROM_SOURCE = Symbol.for("NOT_FROM_SOURCE")
```
Comment (bhavitsharma) [inline ts/private/ts_project_worker.js:282]: Can you explain what files are we going to mark as `NOT_FROM_SOURCE`?

Comment (bhavitsharma) [reply ts/private/ts_project_worker.js:282]: Is it just tsbuildinfo files?

File: ts/private/ts_project_worker.js [hunk 2]
```diff
@@ -279,32 +279,36 @@ function createEmitAndLibCacheAndDiagnosticsProgram(
...
-                emittingMap.set(fileName, sourceFiles[0].fileName);
-                emittedFiles.set(fileName, data);
+        const writeF = (fileName, text, writeByteOrderMark, onError, sourceFiles) => {
+            writeFile(fileName, text, writeByteOrderMark, onError, sourceFiles);
+            outputCache.set(fileName, {text, writeByteOrderMark});
+            if (sourceFiles?.length > 0) {
+                outputSourceMapping.set(fileName, sourceFiles[0].fileName)
+            } else {
+                // if the file write is not the result of a source mark it as not from source not avoid cache drops.
+                outputSourceMapping.set(fileName, NOT_FROM_SOURCE)
```
Comment (bhavitsharma) [inline ts/private/ts_project_worker.js:311]: I guess this is the fix for tsbuildinfo?

Comment (bhavitsharma) [reply ts/private/ts_project_worker.js:311]: OK, I understand the fix now. tsbuildinfo file doesn't have any source, so it's sourceFiles.length will always be zero and hence the bug.

**Review by bhavitsharma — COMMENTED** [review]
LGTM for the PR. Thanks for the fix!

File: ts/private/ts_project_worker.js (hunk 1 above)
Comment (bhavitsharma) [reply ts/private/ts_project_worker.js:282]: I understand the fix now. You can resolve this comment.

File: ts/private/ts_project_worker.js [hunk 3]
```diff
@@ -458,7 +462,7 @@ function createProgram(args, initialInputs) {
     }
 
     function debuglog(message) {
-        compilerOptions.extendedDiagnostics && host.trace?.(message)
```
Comment (alexeagle) [inline ts/private/ts_project_worker.js:461]: do you mean to keep this edit?

**Review by alexeagle — APPROVED** [review]

File: ts/private/ts_project_worker.js (hunk 3 above)
Comment (thesayyn) [reply ts/private/ts_project_worker.js:461]: no. I'll remove that in the next pass.

**Pull request merged**